feat: add flat config support

michaelfaith opened the pull request:
This change adds support for ESLint's new Flat config system.  It maintains backwards compatibility with eslintrc style configs as well.

To achieve this, we're now dynamically creating flat configs on a new `flatConfigs` export.  I was a bit on the fence about using this convention, or the other convention that's become prevalent in the community: adding the flat configs directly to the `configs` object, but with a 'flat/' prefix.  I like this better, since it's slightly more ergonomic when using it in practice.  e.g. `...importX.flatConfigs.recommended` vs `...importX.configs['flat/recommended']`, but i'm open to changing that.

Example Usage

```js
import importPlugin from 'eslint-plugin-import';
import js from '@eslint/js';
import tsParser from '@typescript-eslint/parser';

export default [
  js.configs.recommended,
  importPlugin.flatConfigs.recommended,
  importPlugin.flatConfigs.react,
  importPlugin.flatConfigs.typescript,
  {
    files: ['**/*.{js,mjs,cjs,jsx,mjsx,ts,tsx,mtsx}'],
    languageOptions: {
      parser: tsParser,
      ecmaVersion: 'latest',
      sourceType: 'module',
    },
    ignores: ['eslint.config.js'],
    rules: {
      'no-unused-vars': 'off',
      'import/no-dynamic-require': 'warn',
      'import/no-nodejs-modules': 'warn',
    },
  },
];
```

Closes #29

michaelfaith commented:
Are the unit tests in this repo supposed to be passing?  I noticed they failed in my branch, and so I checked out current `master` branch, and they're also failing for me there.  Is it my setup, or are the just not working?

Zamiell commented on src/config/flat/typescript.ts:
```diff
@@ -0,0 +1,35 @@
+      '@typescript-eslint/parser': [...typeScriptExtensions],
+    },
+    'import-x/resolver': {
+      node: {
```
i think you also need to add "typescript: true" above this line in order for the no-cycle rule to work correctly

Zamiell replied:
(i found this from [this issue](https://github.com/un-ts/eslint-plugin-import-x/issues/29#issuecomment-2119219537))

silverwind replied:
Would doing that not add a dependency on https://github.com/import-js/eslint-import-resolver-typescript?

FWIW, I use only `typescript: true` on my personal config and had no issues linting node scripts so far.

Zamiell replied:
Oh, I was under the impression that `[eslint-import-resolver-typescript](https://github.com/import-js/eslint-import-resolver-typescript)` was mandatory for `eslint-plugin-import-x` + TypeScript. Is that not the case after this PR? If not, ignore my original suggestion.

silverwind replied:
I think `eslint-import-resolver-typescript` is currently only mandatory for typescript, but I think we should make it a dependency and default to it for both JS and TS. The default JS resolver has too many issues like not supporting `exports` object format which are resolved in the typescript resolver.

Zamiell replied:
huge +1 to that. abstract away the complexity for the ned-user, have this plugin "just work"

michaelfaith replied:
Wouldn't this be more appropriate as a separate change?  I was aiming for parity between existing configs and the new flat configs.

SukkaW replied:
> Wouldn't this be more appropriate as a separate change? I was aiming for parity between existing configs and the new flat configs.

Let's do a separate PR!

silverwind commented:
> flat configs on a new flatConfigs export

I like it personally as well as I find `configs['flat/recommended']` a bit ugly.

silverwind commented on src/utils/ignore.ts:
```diff
@@ -28,7 +28,7 @@ function validExtensions(context: ChildContext | RuleContext) {
   // start with explicit JS-parsed extensions
   const exts = new Set<FileExtension>(
-    settings['import-x/extensions'] || ['.js'],
+    settings['import-x/extensions'] || ['.js', '.mjs', '.cjs'],
```
Do we need `jsx` here?

michaelfaith replied:
I don't think so, the `react` specific config adds the .jsx extension.  This is just defining the default / fallback extensions supported, when not using a config or using a config that doesn't define its own.

silverwind commented:
> Are the unit tests in this repo supposed to be passing? I noticed they failed in my branch, and so I checked out current `master` branch, and they're also failing for me there. Is it my setup, or are they just not working?

Tests are passing for me on master branch with `yarn && yarn build && yarn test`:

````
Test Suites: 54 passed, 54 total
Tests:       4 skipped, 2665 passed, 2669 total
Snapshots:   0 total
Time:        7.388 s
````

SukkaW commented:
@michaelfaith We can probably split this PR into two: a PR that adds a flat config preset without `no-unused-modules`, and another PR that adds flat config support for `no-unused-module`.

With this, people can start using the flat presets right away.

michaelfaith commented:
> @michaelfaith We can probably split this PR into two: a PR that adds a flat config preset without `no-unused-modules`, and another PR that adds flat config support for `no-unused-module`.
> 
> With this, people can start using the flat presets right away.

I would be ok doing that, if the maintainers are. I was turned down when i suggested doing that in my PR in the other repo.

SukkaW commented:
> if the maintainers are. I was turned down when i suggested doing that in my PR in the other repo.

It is OK here. I myself am an early flat config adapter (when ESLint 8.40 was released). I'd love to have a flat preset.

As for the inconsistency, we can mention in the changeset that the `no-unused-modules` rule is not available yet.

michaelfaith commented:
> > if the maintainers are. I was turned down when i suggested doing that in my PR in the other repo.
> 
> It is OK here. I myself am an early flat config adapter (when ESLint 8.40 was released). I'd love to have a flat preset.
> 
> As for the inconsistency, we can mention in the changeset that the `no-unused-modules` rule is not available yet.

Sounds good.  Then i'll try and find some time to wrap this up soon.  Thanks

michaelfaith commented:
> > Are the unit tests in this repo supposed to be passing? I noticed they failed in my branch, and so I checked out current `master` branch, and they're also failing for me there. Is it my setup, or are they just not working?
> 
> Tests are passing for me on master branch with `yarn && yarn build && yarn test`:
> 
> ```
> Test Suites: 54 passed, 54 total
> Tests:       4 skipped, 2665 passed, 2669 total
> Snapshots:   0 total
> Time:        7.388 s
> ```

Yeah, that's what I've been running too, but still getting a ton of failures both in my branch and master.  So I'm kind of just relying on CI here since running locally doesn't appear to be reliable. 
![image](https://github.com/user-attachments/assets/2229b533-41f1-4131-97c0-7e59ccb6f6b2)

michaelfaith commented:
This should be ready for review.  

As per the above conversation, this doesn't include addressing issues with `no-unused-module` and can be addressed separately.

> > if the maintainers are. I was turned down when i suggested doing that in my PR in the other repo.
> 
> It is OK here. I myself am an early flat config adapter (when ESLint 8.40 was released). I'd love to have a flat preset.
> 
> As for the inconsistency, we can mention in the changeset that the `no-unused-modules` rule is not available yet.

SukkaW commented on .changeset/great-dodos-dream.md:
```diff
@@ -0,0 +1,5 @@
+"eslint-plugin-import-x": minor
+---
+
+add support for flat configs
```
````suggestion
Add ESLint flat configuration presets. You can access them with:

```ts
import eslintPluginImportX from 'eslint-plugin-import-x';

eslintPluginImportX.flatConfigs.recommended;
eslintPluginImportX.flatConfigs.react;
eslintPluginImportX.flatConfigs.typescript;
eslintPluginImportX.flatConfigs.electron;
```
````
